### ISSUE TITLE
Compile_Using_TransferData: Fail to resume after minimised

### ISSUE THREAD

**Issue by rolandhill** [vsg-dev/VulkanSceneGraph#1277]
Hi Robert,

Just as an FYI, I was using Compile_Using_TransferData commit 80370b9ead50c3ec5302f984298f8e3fee6ee7bf for a while and I noticed that if I minimised my window then un-minimised it again, the window would just be black and I could not close it using the X in the corner. This was running through the debugger and I had to abort it from the debugger. Going back to master branch and all appears fine so far.

Maybe this has already been resolved, but you might want to give it a try and see if you can repeat it.

I'm using Kubuntu 24.04 and NVidia 3070Ti.

Regards,
Roland

**Comment by robertosfield** [comment]
Ahhh.... that commit fixes an issue with latency in ViewDependentState on NVidia which can be seen with vsgshadow.  Just tried opening and closing windows on my Kubuntu 22.04 system with AMD5700G and it works fine.  I'll go try my Windows machine which has my Geforce 2060 in it.

**Comment by robertosfield** [comment]
Does the problem happen with vsgviewer?  i.e

~~~ sh
vsgviewer models/lz.vsgt -d
vsgviewer models/openstreetmap.vsgt -d
~~~

**Comment by robertosfield** [comment]
@rolandhill To try and address the regression you've seen I have rewritten how the wait semaphores are managed in TransferTask, could you try this out? I've check the changes into the Compile_Using_TransferData branch, with commit: adc18e377ffda3653093381c7905ee34f53b1396

If this looks fine for you, and if my testing under Linux and Windows on the machines I have available I'll push ahead and merged with VSG master.

**Comment by robertosfield** [comment]
I've found another regression - when loading models/openstreetmap.vsgt I'm getting validation errors.  This was working last week so in trying to resolve one set of issues, I've tripped up soemthing that was working :-|

**Comment by rolandhill** [comment]
Hi Robert,

I've pulled the latest Compile_Using_TransferData,
commit 171de421f91efd7ad8dc6b172e7d370522264e6f.

When I run my own application in debug mode it starts successfully and
displays a Welcome ImGui dialog and all works as expected. When I select a
recent project to load then the screen freezes and the application
becomes unresponsive, but doesn't crash. What it should be doing is
loading data from disk, compiling it dynamically and then adding branches
into the scene graph. It looks like I'm getting at least one more frame
because another ImGui window appears.

Note that I'm using a fairly complicated setup. One pass writes opaque
objects into the frame buffer, then there is a pipeline barrier and
transparent objects are written to a buffer, then there is a compositor
pass.

The following debug output is shown:

> VUID-vkQueueSubmit-pWaitSemaphores-03238(ERROR / SPEC): msgNum:
> -1258007888 - Validation Error: [ VUID-vkQueueSubmit-pWaitSemaphores-03238
> ] Object 0: handle = 0x44695a0000000071, type = VK_OBJECT_TYPE_SEMAPHORE;
> Object 1: handle = 0x5555573f5260, type = VK_OBJECT_TYPE_QUEUE; | MessageID
> = 0xb50452b0 | vkQueueSubmit(): pSubmits[0].pWaitSemaphores[0] queue
> (VkQueue 0x5555573f5260[]) is waiting on semaphore (VkSemaphore
> 0x44695a0000000071[]) that has no way to be signaled. The Vulkan spec
> states: All elements of the pWaitSemaphores member of all elements of
> pSubmits created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_BINARY must
> reference a semaphore signal operation that has been submitted for
> execution and any semaphore signal operations on which it depends must have
> also been submitted for execution (
> https://vulkan.lunarg.com/doc/view/1.3.280.1/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pWaitSemaphores-03238
> )
>     Objects: 2
>         [0] 0x44695a0000000071, type: 5, name: NULL
>         [1] 0x5555573f5260, type: 4, name: NULL
> VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743
> - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0:
> handle = 0x14ad5800000001a8, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1:
> handle = 0x5555574064b0, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle =
> 0x5555574064b0, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x71565eaf |
> vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] (VkSemaphore
> 0x14ad5800000001a8[]) is being signaled by VkQueue 0x5555574064b0[], but it
> was previously signaled by VkQueue 0x5555574064b0[] and has not since been
> waited on. The Vulkan spec states: Any calls to vkCmdSetEvent,
> vkCmdResetEvent or vkCmdWaitEvents that have been recorded into any of the
> command buffer elements of the pCommandBuffers member of any element of
> pSubmits, must not reference any VkEvent that is referenced by any of those
> commands in a command buffer that has been submitted to another queue and
> is still in the pending state (
> https://vulkan.lunarg.com/doc/view/1.3.280.1/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065
> )
>     Objects: 3
>         [0] 0x14ad5800000001a8, type: 5, name: NULL
>         [1] 0x5555574064b0, type: 4, name: NULL
>         [2] 0x5555574064b0, type: 4, name: NULL



If I pause the debugger, it shows that at least one thread is waiting here:

Fence.cpp
VkResult Fence::wait(uint64_t timeout) const
{
    return vkWaitForFences(*_device, 1, &_vkFence, VK_TRUE, timeout);
}


If I try "vsgviewer ../data/models/lz.vsgt -d", again compiled in debug
mode, the window displays and spins as expected. If I then minimise it then
un-minimise it again I get a blank frozen window and can only regain
control by ^C on the command line.

After un-minimising, the following debug output is shown:

VUID-vkQueueSubmit-pWaitSemaphores-03238(ERROR / SPEC): msgNum: -1258007888
> - Validation Error: [ VUID-vkQueueSubmit-pWaitSemaphores-03238 ] Object 0:
> handle = 0x6871
> 490000000055, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle =
> 0x57f6b0d72f50, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0xb50452b0 |
> vkQueueSubmit(): pSubmits[0]
> .pWaitSemaphores[0] queue (VkQueue 0x57f6b0d72f50[]) is waiting on
> semaphore (VkSemaphore 0x6871490000000055[]) that has no way to be
> signaled. The Vulkan spec states
> : All elements of the pWaitSemaphores member of all elements of pSubmits
> created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_BINARY must reference a
> semaphore signal
> operation that has been submitted for execution and any semaphore signal
> operations on which it depends must have also been submitted for execution (
> https://vulkan.lu
>
> narg.com/doc/view/1.3.280.1/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pWaitSemaphores-03238
> )
>    Objects: 2
>        [0] 0x6871490000000055, type: 5, name: NULL
>        [1] 0x57f6b0d72f50, type: 4, name: NULL
> VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743
> - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0:
> handle = 0x6dcc3
> f000000005b, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle =
> 0x57f6b0d52420, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle =
> 0x57f6b0d52420, type = VK_OBJECT_TYPE
> _QUEUE; | MessageID = 0x71565eaf | vkQueueSubmit():
> pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0x6dcc3f000000005b[]) is
> being signaled by VkQueue 0x57f6b0d52420[],
> but it was previously signaled by VkQueue 0x57f6b0d52420[] and has not
> since been waited on. The Vulkan spec states: Any calls to vkCmdSetEvent,
> vkCmdResetEvent or v
> kCmdWaitEvents that have been recorded into any of the command buffer
> elements of the pCommandBuffers member of any element of pSubmits, must not
> reference any VkEven
> t that is referenced by any of those commands in a command buffer that has
> been submitted to another queue and is still in the pending state (
> https://vulkan.lunarg.co
>
> m/doc/view/1.3.280.1/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065)
>    Objects: 3
>        [0] 0x6dcc3f000000005b, type: 5, name: NULL
>        [1] 0x57f6b0d52420, type: 4, name: NULL
>        [2] 0x57f6b0d52420, type: 4, name: NULL


Regards,
Roland

**Comment by robertosfield** [comment]
Hi Roland,

It sounds like you are seeing a similar problem to me when I run:

~~~ sh
vsgviewer models/opensteetmap.vsgt
~~~

It looks to be down to the paging downloading data in sporadic way and the TransferTask not handling this case properly with the refactor I made over the last few days.  I investigated this regression yesterday afternoon and now have a better handle on why this is happening and have changes that will hopefully address this.

Fingers and toes crosses,
Robert.

**Comment by robertosfield** [comment]
I have refactored how TransferTask manages it's wait for consumer and single transfer semaphores so they can handle sporadic updates, this fixes the regression with paged databases and should also fixes other usage cases where dynamic data is not transferred on all frames.  The fixes are in commit: 0a73a448adab4727826657666b85d7412ebb910e

@rolandhill could you test out the latest head of the Compile_Using_TransferData branch against your application and let me know how you get on.  I will continue testing at my end.

**Comment by rolandhill** [comment]
Hi Robert,

I tested the commit you requested and the results are a bit odd. The
application starts as before and shows the ImGui Welcome window - I tested
4 of my projects that all work with VSG master. If I choose two of the
projects the application locks up as before and generates the same debug
layer messages. If I choose either of the other 2 projects then everything
loads as expected and there are no messages. The scene appears normal and
you can spin it around. With the working projects, if I minimise then
un-minimise, the window is black and the application is unresponsive. In
this case, lots of the following message are displayed:

VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743
> - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0:
> handle = 0x14ad5800000001a8, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1:
> handle = 0x5555574074e0, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle =
> 0x5555574074e0, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x71565eaf |
> vkQueueSubmit(): pSubmits[0].pSignalSemaphores[155] (VkSemaphore
> 0x14ad5800000001a8[]) is being signaled by VkQueue 0x5555574074e0[], but it
> was previously signaled by VkQueue 0x5555574074e0[] and has not since been
> waited on. The Vulkan spec states: Any calls to vkCmdSetEvent,
> vkCmdResetEvent or vkCmdWaitEvents that have been recorded into any of the
> command buffer elements of the pCommandBuffers member of any element of
> pSubmits, must not reference any VkEvent that is referenced by any of those
> commands in a command buffer that has been submitted to another queue and
> is still in the pending state (
> https://vulkan.lunarg.com/doc/view/1.3.280.1/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065
> )
>     Objects: 3
>         [0] 0x14ad5800000001a8, type: 5, name: NULL
>         [1] 0x5555574074e0, type: 4, name: NULL
>         [2] 0x5555574074e0, type: 4, name: NULL
>

I didn't try multiple projects earlier today, so I can't say if some of the
projects worked then.

Just throwing some ideas out there;
One of the projects that doesn't work only consists of a quad-tree
elevation surface with a single image overlay. Admittedly, these are huge
files, but it works well in VSG master. It looks like one to a few frames
get displayed. After the first frame the DatabasePager will be full for a
few seconds as it loads successively higher resolutions. There is a bit of
processing, so it might take a few frames before each request is finalised.
Could a multi-frame delay be causing a problem? You can see a video of what
it should look like here https://www.youtube.com/watch?v=ZlQKZeqgk2g

Regards,
Roland

**Comment by robertosfield** [comment]
Darn.  The changes I've made were intended to address this issue with previously signalled semaphores not being waited upon.  This addressed the Vulkan debug errors I was getting with models/openstreetmap.vsgt, but there must be something else going on as well.

Does `vsgviewer models/openstreetmap.vsgt` work fine for you?

**Comment by rolandhill** [comment]
Both vsgviewer models/openstreetmap.vsgt and vsgviewer models/lz.vsgt work
for me. With both, I minimised then un-minimised them successfully once or
twice, then the next time it resulted in a blank screen and being
unresponsive.

**Comment by robertosfield** [comment]
The "good" news is that I've tried minimizing/unminimizing vsgviewer and get the black window as well under Kubuntu 22.04 with AMD5700G's integrated graphics.  I had just pulled out a Geforce 1650 from storage to start testing with but as I have now recreate the black window issue with the AMD integrated graphics and MESA driver I'll stick with what I have.

I also tested under Windows but I could provoke it to produce any problems despite minimizing/unminizing the window many times.

I don't know the cause of the black window issue, it may or may not be related to the the semaphore not being waited upon validation error.

**Comment by robertosfield** [comment]
I have now identified that cause of failure in the window minimize case, so I'm am contemplating how to refactor things so that the semaphores are better managed in the case where windows are minimized and the usual consumer of transfer data signal semaphores no longer run.

**Comment by robertosfield** [comment]
Hi @rolandhill, I've done several more rounds of revisions to address the synchronization issues and now have things working well for the Windows + Linux with various applications and test models.  Could you test the latest head of Compile_Using_TransferData on your application and models?

**Comment by rolandhill** [comment]
Hi Robert,

Using commit fb6bc6489e5dea8138c0c6731350c9c2ae0cb8d6

*bin/vsgviewer ../data/models/lz.vsgt -d*
Displays as expected and behaving properly minimising/unminimising.
It does throw the following debug message 4 times with different addresses

VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743
> - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0:
> handle = 0xead93
> 70000000008, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle =
> 0x5d3677a47680, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle =
> 0x5d3677a47680, type = VK_OBJECT_TYPE
> _QUEUE; | MessageID = 0x71565eaf | vkQueueSubmit():
> pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0xead9370000000008[]) is
> being signaled by VkQueue 0x5d3677a47680[],
> but it was previously signaled by VkQueue 0x5d3677a47680[] and has not
> since been waited on. The Vulkan spec states: Any calls to vkCmdSetEvent,
> vkCmdResetEvent or v
> kCmdWaitEvents that have been recorded into any of the command buffer
> elements of the pCommandBuffers member of any element of pSubmits, must not
> reference any VkEven
> t that is referenced by any of those commands in a command buffer that has
> been submitted to another queue and is still in the pending state (
> https://vulkan.lunarg.co
>
> m/doc/view/1.3.280.1/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065)
>    Objects: 3
>        [0] 0xead9370000000008, type: 5, name: NULL
>        [1] 0x5d3677a47680, type: 4, name: NULL
>        [2] 0x5d3677a47680, type: 4, name: NULL



*bin/vsgviewer ../data/models/openstreetmap.vsgt -d*
Displays as expected and throws the same debug message as above, but lots
of times (I assume the messages are for each tile). Minimising and
unminimising once produces the same blank, unresponsive window.

*My own application*
Much the same. All my test projects now load and display. The big raster
elevation/image overlay usually works and continues to page and froze just
once. My feeling is that this is an improvement.
Minimising/unminimising results in a blank window and being unresponsive
about 90% of the time. Interestingly, on one run of the big raster project
I could minimise/unminimise around 8 times successfully before I
deliberately closed the application. I tried to do it again and it froze
after a few frames.

This is all on my RTX3070Ti desktop. I'll break out the laptop with a
RTX2070 mobile GPU and test that as well.

Cheers,
Roland

**Comment by rolandhill** [comment]
Just some additional information.

Ubuntu 24.04 on RTX 2070 laptop and RTX 3070Ti desktop - same issues as previously discussed. Both were on NVidia driver 535 from the repos and Vulkan SDK 1.3.280.1. I upgraded to NVidia driver 550 and Vulkan SDK 1.3.130 and rebuilt everything but there was no change.

On the same RTX 2070 laptop in Windows 11, my application ran perfectly. I minimised and unminimised the window multiple times and tried different projects. All was good. I'm only set up to build in release mode, so I don't know if there would have been any debug layer messages.

**Comment by robertosfield** [comment]
THanks for the testing.  I'm seeing Vulkan debug errors with vsgviewer models/openstreetmap.vsgt but not issues with minimizing/opening on my Intel iGPU on my laptop and AMD iGPU on my desktop.

~~~ sh
$ vsgviewer models/openstreetmap.vsgt -d
UNASSIGNED-vkQueueSubmit-QueueForwardProgress(ERROR / SPEC): msgNum: 487709818 - Validation Error: [ UNASSIGNED-vkQueueSubmit-QueueForwardProgress ] Object 0: handle = 0x70000000007, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x1d11dc7a | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0x70000000007[]) is being signaled by VkQueue 0x64b2920ed700[], but it was previously signaled by VkQueue 0x64b2920ed700[] and has not since been waited on
    Objects: 3
        [0] 0x70000000007, type: 5, name: NULL
        [1] 0x64b2920ed700, type: 4, name: NULL
        [2] 0x64b2920ed700, type: 4, name: NULL
UNASSIGNED-vkQueueSubmit-QueueForwardProgress(ERROR / SPEC): msgNum: 487709818 - Validation Error: [ UNASSIGNED-vkQueueSubmit-QueueForwardProgress ] Object 0: handle = 0x70000000007, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x1d11dc7a | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0x70000000007[]) is being signaled by VkQueue 0x64b2920ed700[], but it was previously signaled by VkQueue 0x64b2920ed700[] and has not since been waited on
    Objects: 3
        [0] 0x70000000007, type: 5, name: NULL
        [1] 0x64b2920ed700, type: 4, name: NULL
        [2] 0x64b2920ed700, type: 4, name: NULL
UNASSIGNED-vkQueueSubmit-QueueForwardProgress(ERROR / SPEC): msgNum: 487709818 - Validation Error: [ UNASSIGNED-vkQueueSubmit-QueueForwardProgress ] Object 0: handle = 0x70000000007, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x1d11dc7a | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0x70000000007[]) is being signaled by VkQueue 0x64b2920ed700[], but it was previously signaled by VkQueue 0x64b2920ed700[] and has not since been waited on
    Objects: 3
        [0] 0x70000000007, type: 5, name: NULL
        [1] 0x64b2920ed700, type: 4, name: NULL
        [2] 0x64b2920ed700, type: 4, name: NULL
UNASSIGNED-vkQueueSubmit-QueueForwardProgress(ERROR / SPEC): msgNum: 487709818 - Validation Error: [ UNASSIGNED-vkQueueSubmit-QueueForwardProgress ] Object 0: handle = 0x80000000008, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x1d11dc7a | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0x80000000008[]) is being signaled by VkQueue 0x64b2920ed700[], but it was previously signaled by VkQueue 0x64b2920ed700[] and has not since been waited on
    Objects: 3
        [0] 0x80000000008, type: 5, name: NULL
        [1] 0x64b2920ed700, type: 4, name: NULL
        [2] 0x64b2920ed700, type: 4, name: NULL
UNASSIGNED-vkQueueSubmit-QueueForwardProgress(ERROR / SPEC): msgNum: 487709818 - Validation Error: [ UNASSIGNED-vkQueueSubmit-QueueForwardProgress ] Object 0: handle = 0x80000000008, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x64b2920ed700, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x1d11dc7a | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[1] (VkSemaphore 0x80000000008[]) is being signaled by VkQueue 0x64b2920ed700[], but it was previously signaled by VkQueue 0x64b2920ed700[] and has not since been waited on
    Objects: 3
        [0] 0x80000000008, type: 5, name: NULL
        [1] 0x64b2920ed700, type: 4, name: NULL
        [2] 0x64b2920ed700, type: 4, name: NULL
~~~

**Comment by robertosfield** [comment]
I have spotted an error in the timing of the semaphore rest, this commit 5c342c620df9bcb74327957173bf7a6439d069de fixes the Vulkan debug error I was seeing.  Could you test?

**Comment by rolandhill** [comment]
I'm sorry, Robert. I still see a problem, although it does seem to be better. Different error messages are progress!

I'm using commit 5c342c620df9bcb74327957173bf7a6439d069de

On the big elevation/image overlay I notice that I can consistently successfully minimise/unminimise it if there are no DBPager requests in progress. If I then zoom in to trigger quad tree paging and immediately minimise then I get a blank window and unresponsive when unminimising. The error message has changed too. The following 2 messages are repeated many times

```
VUID-vkResetCommandPool-commandPool-00040(ERROR / SPEC): msgNum: -1254218959 - Validation Error: [ VUID-vkResetCommandPool-commandPool-00040 ] Object 0: handle = 0x5555591c6580, type = VK_OBJECT_TYPE_COMMAND_BUFFER; Object 1: handle = 0x44695a0000000071, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0xb53e2331 | vkResetCommandPool():  (VkCommandBuffer 0x5555591c6580[]) is in use. The Vulkan spec states: All VkCommandBuffer objects allocated from commandPool must not be in the pending state (https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkResetCommandPool-commandPool-00040)
    Objects: 2
        [0] 0x5555591c6580, type: 6, name: NULL
        [1] 0x44695a0000000071, type: 25, name: NULL
VUID-vkBeginCommandBuffer-commandBuffer-00049(ERROR / SPEC): msgNum: -2080204129 - Validation Error: [ VUID-vkBeginCommandBuffer-commandBuffer-00049 ] Object 0: handle = 0x5555591c6580, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x84029a9f | vkBeginCommandBuffer():  on active VkCommandBuffer 0x5555591c6580[] before it has completed. You must check command buffer fence before this call. The Vulkan spec states: commandBuffer must not be in the recording or pending state (https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkBeginCommandBuffer-commandBuffer-00049)
    Objects: 1
        [0] 0x5555591c6580, type: 6, name: NULL

```
Another one of my projects fails to unminimise even if the DBPager queue is empty. It throws the following 3 messages multiple times:

```
VUID-vkResetCommandPool-commandPool-00040(ERROR / SPEC): msgNum: -1254218959 - Validation Error: [ VUID-vkResetCommandPool-commandPool-00040 ] Object 0: handle = 0x55555d5b2ec0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; Object 1: handle = 0x2610a100000001b1, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0xb53e2331 | vkResetCommandPool():  (VkCommandBuffer 0x55555d5b2ec0[]) is in use. The Vulkan spec states: All VkCommandBuffer objects allocated from commandPool must not be in the pending state (https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkResetCommandPool-commandPool-00040)
    Objects: 2
        [0] 0x55555d5b2ec0, type: 6, name: NULL
        [1] 0x2610a100000001b1, type: 25, name: NULL
VUID-vkBeginCommandBuffer-commandBuffer-00049(ERROR / SPEC): msgNum: -2080204129 - Validation Error: [ VUID-vkBeginCommandBuffer-commandBuffer-00049 ] Object 0: handle = 0x55555d5b2ec0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x84029a9f | vkBeginCommandBuffer():  on active VkCommandBuffer 0x55555d5b2ec0[] before it has completed. You must check command buffer fence before this call. The Vulkan spec states: commandBuffer must not be in the recording or pending state (https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkBeginCommandBuffer-commandBuffer-00049)
    Objects: 1
        [0] 0x55555d5b2ec0, type: 6, name: NULL
VUID-vkQueueSubmit-pCommandBuffers-00071(ERROR / SPEC): msgNum: 774851941 - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00071 ] | MessageID = 0x2e2f4d65 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] VkCommandBuffer 0x55555d5b2ec0[] is already in use and is not marked for simultaneous use. The Vulkan spec states: If any element of the pCommandBuffers member of any element of pSubmits was not recorded with the VK_COMMAND_BUFFER_USAGE_SIMULTANEOUS_USE_BIT, it must not be in the pending state (https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00071)
    Objects: 0

```
The first 2 are the same as with the large raster project it appears.

**bin/vsgviewer ../data/models/openstreetmap.vsgt -d**
This does the same as the large raster in the first example. I can start it, minimise/unminimise all OK. If I zoom in to trigger database paging then minimise, it will blank screen and be unresponsive. Try zooming in then minimising there to see if you can replicate it.

Cheers,
Roland

**Comment by robertosfield** [comment]
Different errors, I guess that's progress?!?# :-)

I have tried vsgviewer with openstreetmap.vsgt but on the Kubuntu 24.04 Intel laptop I can't get it to produce any issues with minizing/unminizing the window.  I don't have access to my desktop system right now so can't test this.  

I may need to install a NVidia graphics card on my desktop system to investigate whether it's now an issue that only the NVidia cards/drivers can reveal.  What drivers are you using?

It would be useful to know which command buffer reset is causing the error, my guess is it would be one in TransferData but this should all be protected by the fence in RecordAndSubmitTask.  Could you run with Vulkan API debug layer on and find the calls immediately prior to the validation error, hopefully this will give us a clue where to look.

FYI, with the Compile_Using_TransferData branch the DatabasePager's compile traversal will create Vulkan objects and allocate memory but won't do the copy to the GPU, by default that's now done by TransferTask.  This change avoids the need to record and submit a command buffer within the the CompileTraversal, and crucially avoids the wait in the DatabasePager as there is nothing to wait for.  Instead TransferTask will do the copy and use a semaphore to coordinate the write with the final rendering.  

This change substantially lowers latency in the DatabasePager so it's now able to load and get data on screen faster, which increases the resolution of the LOD's that are able to be displayed for a given time.  This in turn puts more load the deleting old PagedLOD subgraphs, which is why there is now vsg::DeleteQueue and a dedicated delete thread in DatabasePager.  Without DeleteQueue and delete thread I was getting lots of big stalls once the number of PagedLOD hit MaxPagedLOD and expiry of old subgraphs kicked in.

It might be that the new rate at which new data can loaded and old data released is stressing things much more.  Once all the teething issues are resolved which should have better paged databases experience with requested tiles being loaded and rendered with lower latency.  Unfortunately, these problem are proving hard to resolve :-|

**Comment by rolandhill** [comment]
I'm using NVidia proprietary driver 550 from the Kubuntu 24.04 repos.

I've attached a log file generated with windowTraits debugLayer and
apiDumpLayer both true. Is this what you needed?

**Comment by robertosfield** [comment]
I can't see the attachments, either on the email I received on looking at the Issue tracker. If you can't attach them could just email them to me directly. Thanks.

**Comment by rolandhill** [comment]
Log file sent by email.

Cheers,
Roland

**Comment by robertosfield** [comment]
Thanks, got file.  I've done a first pass review.  

I haven't seen validation errors reported about the command pool, these errors suggest that they execution of the TransferTask::transferData() is happening without waiting for the main frame fence, but that fence should be waited on by RecordAndTransferTask::start() prior to the early and late transfers are done.  

I do wonder if there might be an issue lurking when the window is minimized and main rendering is disabled so that the normal fence that is waited upon isn't done.

What are the circumstances when you see these errors?  Are you running the viewer multi-threaded or with multiple views running?

**Comment by rolandhill** [comment]
Hi Robert,

There is only one view.

I do call setupThreading() during VSG setup, but noticed yesterday that
threading was set to false. At the time I was calling it too early when
there were no commandGraphs installed. I've now moved it to just before the
initial compile() and see that threading is working although it doesn't
appear to make any difference. If I run through the debugger,
unminimise and get a blank window, then pause the application and look
at the call stack, I see different results depending on whether threading
is true or false:

*Threading: false*
Pausing after a failure shows that one thread is sitting at
vsg::Fence::wait() vkWaitForFences(). This is called
from RecordAndSubmitTask::start()  (see annotations below)
VkResult Fence::wait(uint64_t timeout) const
{
return vkWaitForFences(*_device, 1, &_vkFence, VK_TRUE, timeout); <-------
waiting here
}


VkResult RecordAndSubmitTask::start()
{
CPU_INSTRUMENTATION_L1_NC(instrumentation, "RecordAndSubmitTask start",
COLOR_RECORD);

earlyDataTransferred = false;
lateDataTransferred = false;

auto current_fence = fence();
if (current_fence->hasDependencies())
{
// info("RecordAndSubmitTask::start() waiting on fence ", current_fence, ",
", current_fence->status(), ", current_fence->hasDependencies() = ",
current_fence->hasDependencies());

uint64_t timeout = std::numeric_limits<uint64_t>::max();
if (VkResult result = current_fence->wait(timeout); result != VK_SUCCESS)
return result; <------- called from here

current_fence->resetFenceAndDependencies();

// info("after RecordAndSubmitTask::start() waited on fence ",
current_fence, ", ", current_fence->status(), ",
current_fence->hasDependencies() = ", current_fence->hasDependencies());
}
else
{
// info("RecordAndSubmitTask::start() initial fence ", current_fence, ", ",
current_fence->status(), ", current_fence->hasDependencies() = ",
current_fence->hasDependencies());
}

return VK_SUCCESS;
}

Inspecting _device in vkWaitForFences shows the following
[image: image.png]



*Threading: true*
This shows one thread waiting at a Barrier (see annotations below). Note
that I do install a Barrier between render passes, but if I remove them it
still waits here.
class Barrier : public Inherit<Object, Barrier>
{
public:
explicit Barrier(uint32_t num_thread) :
_num_threads(num_thread),
_num_arrived(0),
_phase(0) {}

Barrier(const Barrier&) = delete;
Barrier& operator=(const Barrier&) = delete;

/// increment the arrived count and release the barrier if count matches
number of threads to arrive otherwise wait for the arrived count to match
the number of threads to arrive
void arrive_and_wait()
{
std::unique_lock lock(_mutex);
if (++_num_arrived == _num_threads)
{
_release();
}
else
{
auto my_phase = _phase;
_cv.wait(lock, [this, my_phase]() { return this->_phase != my_phase; });
<------ waiting here
}
}


_num_threads in the above is 2. _num_arrived is 1, so this must be the
first one because it has already incremented.
void Viewer::recordAndSubmit()
{
CPU_INSTRUMENTATION_L1_NC(instrumentation, "Viewer recordAndSubmitTask",
COLOR_VIEWER);

// reset connected ExecuteCommands
for (auto& recordAndSubmitTask : recordAndSubmitTasks)
{
for (auto& commandGraph : recordAndSubmitTask->commandGraphs)
{
commandGraph->reset();
}
}

#if 1
if (_threading)
#else
// The following is a workaround for an odd "Possible data race during
write of size 1" warning that valgrind tool=helgrind reports
// on the first call to vkBeginCommandBuffer despite them being done on
independent command buffers. This could well be a driver bug or a false
positive.
// If you want to quieten this warning then change the #if above to #if 0
as rendering the first three frames single threaded avoids the warning.
if (_threading && _frameStamp->frameCount > 2)
#endif
{
_frameBlock->set(_frameStamp);
_submissionCompleted->arrive_and_wait(); <--------- called from here
}
else
{
for (auto& recordAndSubmitTask : recordAndSubmitTasks)
{
recordAndSubmitTask->submit(_frameStamp);
}
}
}


You said "...  when the window is minimized and main rendering is disabled
...". It doesn't look like rendering is disabled when the window is
minimised. When I start an application (that is destined to fail) it shows
the scene perfectly and there are no debug layer messages. When I minimise
the window the error messages start and continue to flow until I
unminimise it and get the blank screen, at which point the messages stop.
At this point the application appears to be frozen waiting for a fence or
CV (depending on threading). The following 2 messages are shown repeatedly
WHILE the window is minimised

VUID-vkResetCommandPool-commandPool-00040(ERROR / SPEC): msgNum:
> -1254218959 - Validation Error: [ VUID-vkResetCommandPool-commandPool-00040
> ] Object 0: handle = 0x7fffa4049c20, type = VK_OBJECT_TYPE_COMMAND_BUFFER;
> Object 1: handle = 0x36bd8200000001b2, type = VK_OBJECT_TYPE_COMMAND_POOL;
> | MessageID = 0xb53e2331 | vkResetCommandPool():  (VkCommandBuffer
> 0x7fffa4049c20[]) is in use. The Vulkan spec states: All VkCommandBuffer
> objects allocated from commandPool must not be in the pending state (
> https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkResetCommandPool-commandPool-00040
> )
>     Objects: 2
>         [0] 0x7fffa4049c20, type: 6, name: NULL
>         [1] 0x36bd8200000001b2, type: 25, name: NULL
> VUID-vkBeginCommandBuffer-commandBuffer-00049(ERROR / SPEC): msgNum:
> -2080204129 - Validation Error: [
> VUID-vkBeginCommandBuffer-commandBuffer-00049 ] Object 0: handle =
> 0x7fffa4049c20, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID =
> 0x84029a9f | vkBeginCommandBuffer():  on active VkCommandBuffer
> 0x7fffa4049c20[] before it has completed. You must check command buffer
> fence before this call. The Vulkan spec states: commandBuffer must not be
> in the recording or pending state (
> https://vulkan.lunarg.com/doc/view/1.3.290.0/linux/1.3-extensions/vkspec.html#VUID-vkBeginCommandBuffer-commandBuffer-00049
> )
>     Objects: 1
>         [0] 0x7fffa4049c20, type: 6, name: NULL


The only consistent trigger for this problem that I have been able to find
is that if I let the DatabasePager empty its queue then I can
minimise/unminimise with no problems and no errors. If the DatabasePager is
active then it starts throwing errors when the window is minimised. I
removed all my DBPager code and just went with the stock vsg::DatabasePager
installed by default during the initial compile() (enabled by adding an
empty PagedLOD to the scene, because all my content is added dynamically
after the initial compile()). This made no difference.

I also removed all my code to render transparency as a separate pass and
all the Barriers so that there was just a single simple render pass and it
made no difference.

I hope this helps.

Regards,
Roland

**Comment by robertosfield** [comment]
Thanks for the info.  

Unfortunately over the weekend my main development machine has decided to stop booting.  I've unplugged tested, put back in and tested all the components I can from the system and it doesn't boot so it looks like either the GPU or the motherboard has failed.  

This means I'll need to figure out what components I need to buy, order them, get them installed and then try again.  All  extra work I could do without.  I have a working Windows 11 machine and this Kubuntu 24.04 laptop but neither is my go to for best productivity, but at least I can get on a bit.

**Comment by robertosfield** [comment]
I realized I hadn't tried changing the 3V coin battery on the motherboard so I've changed this and now my machine boots, albeit without the original bios settings.

**Comment by rolandhill** [comment]
Well, that's a relief.

Here's my tip - make sure you choose the correct memory profile when
setting up your BIOS. Mine changed after a BIOS update and I spent weeks
trying to figure out why some applications were crashing. I was just about
to buy a replacement 64GB and checked the timings in the BIOS and noticed
that they didn't look right. Major expenditure averted :)

Let me know if I can test anything else.

Cheers,
Roland

**Comment by robertosfield** [comment]
I noticed issues with the memory setup after replacing the battery - only 1 memory module was working and at a low speed.  After pulling out and trying various memory combinations it's all back to where it was originally and with the bios fixed it's all back working with full memory at full speed.

I plan to upgrade this machine later in the year to get an AMD chip with onboard ray tracing support, but doing it today wasn't in the plan so it's a relief to be back working again without any more downtime.

**Comment by robertosfield** [comment]
@rolandhill I have added a vsg::Fence to the TransferTask::TransferBlock to ensure that associated command buffer isn't rerecorded while the previous submission is still outstanding.  Could you test this on your hardware/application?

    b245c1a03aae69bf046ac96d986f5249eade6a19

As part of this commit I have also added debug output to Xcb_Window.cpp to report when window is mapped and unmapped so we have a bit of anchor for the timing of any further Vulkan validation errors that might still be lurking.

**Comment by rolandhill** [comment]
Using commit b245c1a03aae69bf046ac96d986f5249eade6a19

That appears to have made a difference and there are less error messages and they are different. To the end user there is no change with the window being blank and unresponsive if the DatabasePager is busy, while it appears to work as expected if you let the pager queue clear before minimising.

I've attached 2 logs:

- ClearDBQueue just shows your new debug messages as I minimise/unminimise the window.

- busyDBQueue shows messages when I minimise while the pager queue is processing. This results in the blank window. The other notable difference is that when I minimise there are just a few debug messages, then a few more when unminimising. In the past there was a continuous stream of debug messages while minimised.

[clearDBQueue.txt](https://github.com/user-attachments/files/17021033/clearDBQueue.txt)
[busyDBQueue.txt](https://github.com/user-attachments/files/17021035/busyDBQueue.txt)

Cheers,
Roland

**Comment by robertosfield** [comment]
When using:

~~~ sh
 vsgviewer models/openstreetmap.vsgt -p saved_animation.vsgt -d 
~~~

With a saved_animation.vsgt pre-recorded to follow a path that results in lots of paging I am no longer seeing any Vulkan debug errors when I minimize and map the window, and don't see any blank window.  It all just works solidly.  I also added a --mt option to vsgviewer to make it easy to enable the viewer multi-threading, but this doesn't change any w.r.t behavior or Vulkan validation errors, it's all behaving OK.

However, this is on a Kubuntu 22.04 + AMD5700G system, not with a NVidia graphics card so it looks like I'll need to install one and see if I can see any differences.  Do you have any non NVidia systems?

**Comment by robertosfield** [comment]
New day, old problems... my main development desktop is refusing to boot again, my best guess is there is an issue with the motherboard so I'll need to research what hardware to get next this will mean more downtime unfortunately.

I have booted my Windows 11 desktop and will test the Compile_Using_TransferTask branch.  This Windows 11 systems has an iCore7 + Geforce 2060 is closer to what you are using but obviously the OS is quite different.

**Comment by robertosfield** [comment]
@rolandhill I haven't been able to boot my main Linux desktop so have ended up having to spend the morning figuring out what spec of components to upgrade this broken machine to.  An AMD8700G + motherboard + memory should be with me tomorrow, this will all need assembling and OS/dev tools installing etc. so more downtime not fixing bugs...

While juggling the above this morning I got the VSG Compile_Using_TransferData branch updated to the latest  on my Intel/Geforce 2060/Windows 11 machine and have been testing vsgtiledatabase and everything looks fine most of the time but occasionally I've been able get a cluster of the following Vulkan debug errors:

~~~ sh
> vsgtiledatabase.exe -d
VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743 - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0: handle = 0x5c5283000000003e, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x2579e8c6860, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x2579e8c6860, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x71565eaf | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[0] (VkSemaphore 0x5c5283000000003e[]) is being signaled by VkQueue 0x2579e8c6860[], but it was previously signaled by VkQueue 0x2579e8c6860[] and has not since been waited on. The Vulkan spec states: Any calls to vkCmdSetEvent, vkCmdResetEvent or vkCmdWaitEvents that have been recorded into any of the command buffer elements of the pCommandBuffers member of any element of pSubmits, must not reference any VkEvent that is referenced by any of those commands in a command buffer that has been submitted to another queue and is still in the pending state (https://vulkan.lunarg.com/doc/view/1.3.283.0/windows/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065)
    Objects: 3
        [0] 0x5c5283000000003e, type: 5, name: NULL
        [1] 0x2579e8c6860, type: 4, name: NULL
        [2] 0x2579e8c6860, type: 4, name: NULL
VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743 - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0: handle = 0x5c5283000000003e, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x2579e8c6860, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x2579e8c6860, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x71565eaf | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[0] (VkSemaphore 0x5c5283000000003e[]) is being signaled by VkQueue 0x2579e8c6860[], but it was previously signaled by VkQueue 0x2579e8c6860[] and has not since been waited on. The Vulkan spec states: Any calls to vkCmdSetEvent, vkCmdResetEvent or vkCmdWaitEvents that have been recorded into any of the command buffer elements of the pCommandBuffers member of any element of pSubmits, must not reference any VkEvent that is referenced by any of those commands in a command buffer that has been submitted to another queue and is still in the pending state (https://vulkan.lunarg.com/doc/view/1.3.283.0/windows/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065)
    Objects: 3
        [0] 0x5c5283000000003e, type: 5, name: NULL
        [1] 0x2579e8c6860, type: 4, name: NULL
        [2] 0x2579e8c6860, type: 4, name: NULL
VUID-vkQueueSubmit-pWaitSemaphores-03238(ERROR / SPEC): msgNum: -1258007888 - Validation Error: [ VUID-vkQueueSubmit-pWaitSemaphores-03238 ] Object 0: handle = 0x5c5283000000003e, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x2579e777f70, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0xb50452b0 | vkQueueSubmit(): pSubmits[0].pWaitSemaphores[1] queue (VkQueue 0x2579e777f70[]) is waiting on semaphore (VkSemaphore 0x5c5283000000003e[]) that has no way to be signaled. The Vulkan spec states: All elements of the pWaitSemaphores member of all elements of pSubmits created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_BINARY must reference a semaphore signal operation that has been submitted for execution and any semaphore signal operations on which it depends must have also been submitted for execution (https://vulkan.lunarg.com/doc/view/1.3.283.0/windows/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pWaitSemaphores-03238)
    Objects: 2
        [0] 0x5c5283000000003e, type: 5, name: NULL
        [1] 0x2579e777f70, type: 4, name: NULL
VUID-vkQueueSubmit-pWaitSemaphores-03238(ERROR / SPEC): msgNum: -1258007888 - Validation Error: [ VUID-vkQueueSubmit-pWaitSemaphores-03238 ] Object 0: handle = 0x5c5283000000003e, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x2579e777f70, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0xb50452b0 | vkQueueSubmit(): pSubmits[0].pWaitSemaphores[2] queue (VkQueue 0x2579e777f70[]) is waiting on semaphore (VkSemaphore 0x5c5283000000003e[]) that has no way to be signaled. The Vulkan spec states: All elements of the pWaitSemaphores member of all elements of pSubmits created with a VkSemaphoreType of VK_SEMAPHORE_TYPE_BINARY must reference a semaphore signal operation that has been submitted for execution and any semaphore signal operations on which it depends must have also been submitted for execution (https://vulkan.lunarg.com/doc/view/1.3.283.0/windows/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pWaitSemaphores-03238)
    Objects: 2
        [0] 0x5c5283000000003e, type: 5, name: NULL
        [1] 0x2579e777f70, type: 4, name: NULL
~~~

**Comment by robertosfield** [comment]
Looking at my errors on my Windows 11 + Geforce 2060 system and your busyDBQueue.txt they look the same to me.  My current thought is that the transfer task is submitting it's command buffer to do the copy of data to the GPU, the submission has a signal semaphore that is usually used by the RecordAndSubmitTask's submissions as a wait semaphore, but in the case when the window is minimized this wait never happens so the signal is orphaned.

I earlier modified the RecordAndSubmitTask::finish() method to only assign the wait semaphore back to the TransferTask, this addressed part of the problem with opening/closing windows but it doesn't reset the orphaned TransferTask semaphore that is no longer required. 

I will have a ponder how we might resolve this.

**Comment by robertosfield** [comment]
I have added some debug output code to RecordAndSubmitTask::finish() which gets reported when there is no command buffers recorded (which happens when the window is minimized) and when I minimize the window I get a whole bunch of frames where the TransferTask has submitted copies to a queue and signal semaphores from it don't get matched to a main rendering submissions so are essentially orphaned with no-one to wait of them. Eventually I start seeing Vulkan Validation ERRORS report so the output together looks like:

~~~ sh
info: RecordAndSubmitTask::finish() recordedCommandBuffers->empty(), transientWaitSemaphores.size() = 1
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
info: RecordAndSubmitTask::finish() recordedCommandBuffers->empty(), transientWaitSemaphores.size() = 1
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743 - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0: handle = 0x5c5283000000003e, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x1b6e6bdef30, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x1b6e6bdef30, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x71565eaf | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[0] (VkSemaphore 0x5c5283000000003e[]) is being signaled by VkQueue 0x1b6e6bdef30[], but it was previously signaled by VkQueue 0x1b6e6bdef30[] and has not since been waited on. The Vulkan spec states: Any calls to vkCmdSetEvent, vkCmdResetEvent or vkCmdWaitEvents that have been recorded into any of the command buffer elements of the pCommandBuffers member of any element of pSubmits, must not reference any VkEvent that is referenced by any of those commands in a command buffer that has been submitted to another queue and is still in the pending state (https://vulkan.lunarg.com/doc/view/1.3.283.0/windows/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065)
    Objects: 3
        [0] 0x5c5283000000003e, type: 5, name: NULL
        [1] 0x1b6e6bdef30, type: 4, name: NULL
        [2] 0x1b6e6bdef30, type: 4, name: NULL
info: RecordAndSubmitTask::finish() recordedCommandBuffers->empty(), transientWaitSemaphores.size() = 2
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
VUID-vkQueueSubmit-pCommandBuffers-00065(ERROR / SPEC): msgNum: 1901485743 - Validation Error: [ VUID-vkQueueSubmit-pCommandBuffers-00065 ] Object 0: handle = 0x5c5283000000003e, type = VK_OBJECT_TYPE_SEMAPHORE; Object 1: handle = 0x1b6e6bdef30, type = VK_OBJECT_TYPE_QUEUE; Object 2: handle = 0x1b6e6bdef30, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x71565eaf | vkQueueSubmit(): pSubmits[0].pSignalSemaphores[0] (VkSemaphore 0x5c5283000000003e[]) is being signaled by VkQueue 0x1b6e6bdef30[], but it was previously signaled by VkQueue 0x1b6e6bdef30[] and has not since been waited on. The Vulkan spec states: Any calls to vkCmdSetEvent, vkCmdResetEvent or vkCmdWaitEvents that have been recorded into any of the command buffer elements of the pCommandBuffers member of any element of pSubmits, must not reference any VkEvent that is referenced by any of those commands in a command buffer that has been submitted to another queue and is still in the pending state (https://vulkan.lunarg.com/doc/view/1.3.283.0/windows/1.3-extensions/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00065)
    Objects: 3
        [0] 0x5c5283000000003e, type: 5, name: NULL
        [1] 0x1b6e6bdef30, type: 4, name: NULL
        [2] 0x1b6e6bdef30, type: 4, name: NULL
info: RecordAndSubmitTask::finish() recordedCommandBuffers->empty(), transientWaitSemaphores.size() = 3
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
info:    semaphore = ref_ptr<vsg::Semaphore>(vsg::Semaphore 000001B6DF7B2714), semaphore->vk() = 5C5283000000003E
info: RecordAndSubmitTask::finish() recordedCommandBuffers->empty(), transientWaitSemaphores.size() = 3
~~~ sh

There looks to be two issues here - the orphaned semaphores and the fact that these orphaned semaphores just back up, even the same semaphore can be queued up waiting to be waited upon!

As binary semaphores can't be reset other than being waited upon by a submit I'm now wondering if RecordAndSubmitTask::finish() could assign the signal semaphore back to the TransferTask to get it to wait on it the next time it wants to submit some copy work.

**Comment by robertosfield** [comment]
@rolandhill I have come up with a provisional fix for the problem of semaphores getting orphaned when a window is closed and the DatabasePager is still feeding data to be transferred to TransferTask: 3798d5fb43d05bb52f4c04a6ab93e4921f510ac3

This fix resolves the Vulkan Validation Layer ERRORS that I was seeing when using vsgtiledatabase -d and then pressing the various 1 to 8 location keys then immediately closing the window before the animation has completed moving to the next location.

Could you test the Compile_Using_TransferTask on your application and data and let me know how you get on.  If this works I'll clean up the code and make this branch ready to merge with VSG master.

**Comment by rolandhill** [comment]
Success!!!

I can now zoom around minimise/unminimise/maximise on my Ubuntu desktop with a RTX3070Ti with no problems and no debug messages.

It also all works on my Ubuntu RTX2070 laptop, although I did get a single message below. 
```
UNASSIGNED-Threading-MultipleThreads-Write(ERROR / SPEC): msgNum: -1604639890 - Validation Error: [ UNASSIGNED-Threading-MultipleThreads-Write ] Object 0: handle = 0x5555574e1470, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0xa05b236e | vkQueueSubmit():  THREADING ERROR : object of type VkQueue is simultaneously used in current thread 140736649496256 and thread 140736659982016
    Objects: 1
        [0] 0x5555574e1470, type: 4, name: NULL

```

The same RTX2070 laptop dual boots to Windows 11. I tested there too and everything worked as expected with no debug layer messages.

That's all the test platforms I have access to.

Thanks for all your hard work, Robert!

Cheers,
Roland

**Comment by robertosfield** [comment]
Wahoooo!!! :-)

Thanks for all the testing.  The multi-thread Vulkan validation error is one we've seen pop up now and then with never any indication on the VSG side what might be causing it.  The VSG uses a mutex to protect calls to in vsg::Queue so it's hard to know what opportunity there is for such an issue.

**Comment by robertosfield** [comment]
I will now do a code review and clean up, if all looks good will merge with VSG master this morning and get a dev release tagged.

**Comment by robertosfield** [comment]
I have done my code review and clean up and have merged VSG master: 1279.

Could you test out VSG, vsgXchange and vsgExamples master with your app and data?  I'm ready to tag dev releases if everything looks solid.

**Comment by rolandhill** [comment]
Hi Robert,

I've updated to master for the 3 libraries and tested in debug and release
modes on Ubuntu 24.04 RTX3070Ti. Everything looks solid to me. No issues.

Cheers,
Roland

**Comment by robertosfield** [comment]
Great, thanks for the testing and feedback.  I'll now close this thread :-)
